content/docs/specs/native_histograms.md

```diff
@@ -0,0 +1,2353 @@
+section](#zero-bucket) below.
+
+The current limits of -4 for the lowest resolution and 8 for the highest
+resolution have been chosen based on practical usefulness. Should a practical
+need arise for even lower or higher resolution, an extension of the range will
+be considered. However, a schema greater than 52 does not make sense as the
+growth factor from one bucket to the next would then be smaller than the
+difference between representable float64 numbers. Likewise, a schema smaller
+than -9 does not make sense either, as the growth factor would then exceed the
+largest float representable as float64. Therefore, the schema numbers between
+(and including) -9 and +52 are reserved for future standard schemas (following
+the formulas for bucket boundaries above) and MUST NOT be used for any other
+schemas.
+
+For schema -53, the bucket boundaries are set explicitly via _custom values_,
+described in detail in the [custom values section](#custom-values) below. This
+results in a native histogram with custom bucket boundaries (or short _custom
+buckets_, often further abbreviated to NHCB). Such a histogram can be used to
+represent a classic histogram as a native histogram. It can also be used if
+the exponential bucketing featured by the standard schemas is a bad match for
+the distribution to be represented by the histogram. Histograms with different
+custom bucket boundaries are generally not mergeable with each other.
+Therefore, schema -53 SHOULD only be used as an informed decision in specific
+use cases. (TODO: NHCB aren't fully merged into main as of now (2024-11-03).
+They are worked into this document as far as possible already. This information
+might not yet be relevant for released Prometheus versions.)
+
+### Buckets
+
+For standard schemas, buckets are represented as two lists, one for positive
+buckets and one for negative buckets. For custom buckets (schema -53), only the
+positive bucket list is used, but repurposed for all buckets.
+
+Any unpopulated buckets MAY be excluded from the lists. (Which is the reason
+why the buckets are often called _sparse buckets_.)
+
+For float histograms, the elements of the lists are float64 and
+represent the bucket population directly.
+
+For integer histograms, the elements of the lists are signed 64-bit integers
+(short: int64), and each element represents the bucket population as a delta to
+the previous bucket in the list. The first bucket in each list contains an
+absolute population (which can also be seen as a delta relative to zero).
+
+To map buckets in the lists to the indices as defined in the previous section,
+there are two lists of so-called _spans_, one for the positive buckets and one
+for the negative buckets.
+
+Each span consists of a pair of numbers, a signed 32-bit integer (short: int32)
+called _offset_ and an unsigned 32-bit integer (short: uint32) called _length_.
+Only the first span in each list can have a negative offset. It defines the
```

Comment:
For NHCB not even the first offset can be negative as the offset is interpreted as an offset into the custom values.